README.md
[![Auth0](https://cldup.com/yYelLVuXby.jpg)](http://auth0.com)

# Lock
[![NPM version][npm-image]][npm-url]
[![Build status][strider-image]][strider-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![Dependency Status][david-image]][david-url]
[![License][license-image]][license-url]
[![Downloads][downloads-image]][downloads-url]

[Auth0](https://auth0.com) is an authentication broker that supports social identity providers as well as enterprise identity providers such as Active Directory, LDAP, Google Apps, Salesforce.

Lock makes it easy to integrate SSO in your app. You won't have to worry about:

* Having a professional looking login dialog that displays well on any resolution and device.
* Finding the right icons for popular social providers.
* Remembering what was the identity provider the user chose the last time.
* Solving the home realm discovery challenge with enterprise users (i.e.: asking the enterprise user the email, and redirecting to the right enterprise identity provider).
* Implementing a standard sign in protocol (OpenID Connect / OAuth2 Login)

[![Auth0](https://i.cloudup.com/6opoEX_Z9z.png)](http://auth0.com)

> You can try it out yourself online at our [Auth0 Lock playground][playground-url].

## Install

From [npm](https://npmjs.org):

```sh
npm install auth0-lock
```

From [bower](http://bower.io):

```sh
bower install auth0-lock
```

Or our CDN:

```html
<!-- Latest major release -->
<script src="http://cdn.auth0.com/js/lock-6.min.js"></script>

<!-- Latest minor release -->
<script src="http://cdn.auth0.com/js/lock-6.x.min.js"></script>

<!-- Latest patch release (recommended for production) -->
<script src="http://cdn.auth0.com/js/lock-6.x.y.min.js"></script>
```

If you are targeting mobile audiences, it's recommended that you add:

```html
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=0"/>
```

### Browserify

If you are using browserify to build your project, you will need to add the following transformations required by Auth0 Lock:

``` json
{
  "devDependencies": {
    "brfs": "0.0.8",
    "ejsify": "0.1.0",
    "packageify": "^0.2.0"
  }
}
```


## Usage

You can use **Auth0Lock** with [Popup mode][popup-mode] or [Redirect mode][redirect-mode]. To learn more about these modes, you can read the [Authentication Modes][authentication-modes] page. 
There are different ways of implementing them according to your application needs. To see what kind of settings you should be using you can check the [Types of Applications article][application-types].

```js
// Initialize Auth0Lock with your `clientID` and `domain`
var lock = new Auth0Lock('xxxxxx', '<account>.auth0.com');

// and deploy it
var login = document.querySelector('a#login')

login.onclick = function (e) {
  e.preventDefault();
  lock.show(function onLogin(err, profile, id_token) {
    if (err) {
      // There was an error logging the user in
      return alert(err.message);
    }

    // User is logged in
  });
};
```

This is just one example of how **Auth0Lock** could work with a **Single Page Application** (_SPA_). Read the [Single Page Applications][spa-notes] and the [Regular Web Applications][webapps-notes] articles for a full explanation on how to implement those scenarios with Auth0 Lock and when to use each.


## API

### Auth0Lock(clientID, domain[, options])

Initialize `Auth0Lock` with a `clientID` and the account's `domain`.

```js
var lock = new Auth0Lock('xxxxxx', '<account>.auth0.com');
```

> Note: For a full detail on options and parameters you can check the [Auth0Lock initialization][lock-initialization] wiki notes.


### .show([options, callback])

Open the widget on `signin` mode with `signup` and `reset` button actions if enabled for the configured/default account connection.

You may call this method with a single parameter, two or even none. The following examples ilustrate this:

```js
var lock = new Auth0Lock('xxxxxx', '<account>.auth0.com');

// using defaults and resolved configuration
// from you account details
lock.show();

// passing some options
lock.show(options);

// provide with a callback `fn` to be invoked
// at success or error authentication
lock.show(function (err, profile, token) {});

// or both options and callback
lock.show(options, function (err, profile, token) {});

```

> Check the [Auth0Lock customization][lock-customization] article for more examples and options specification. Or enter the [Authentication modes][application-types] notes to learn more about implementing different authentication mechanics.

### .showSignin([options, callback])

Open the widget on `signin` mode, but withouht the bottom `signup` nor `reset` button actions. This method is useful when your site has custom *signup* and *reset* links at a different form.

You may call this method with a single parameter, two or even none. The following examples ilustrate this:

```js
var lock = new Auth0Lock('xxxxxx', '<account>.auth0.com');

// using defaults and resolved configuration
// from you account details
lock.showSignin();

// passing some options
lock.showSignin(options);

// provide with a callback `fn` to be invoked
// on `reset` success or error
lock.showSignin(function (err, profile, token) {});

// or both options and callback
lock.showSignin(options, function (err, profile, token) {});
```
> Check the [Auth0Lock customization][lock-customization] article for more examples and options specification. Or enter the [Authentication modes][application-types] notes to learn more about implementing different authentication mechanics.


### .showSignup([options, callback])

Open the widget on `signup` mode, but withouht the `cancel` button action to go back to `signin`. This method is useful when your site has custom *signin* and *reset* links at a different form.

You may call this method with a single parameter, two or even none. The following examples ilustrate this:

```js
var lock = new Auth0Lock('xxxxxx', '<account>.auth0.com');

// using defaults and resolved configuration
// from you account details
lock.showSignup();

// passing some options
lock.showSignup(options);

// provide with a callback `fn` to be invoked
// on `reset` success or error
lock.showSignup(function (err) {});

// or both options and callback
lock.showSignup(options, function (err) {});
```
> Check the [Auth0Lock customization][lock-customization] article for more examples and options specification. Or enter the [Authentication modes][application-types] notes to learn more about implementing different authentication mechanics.


### .showReset([options, callback])

Open the widget on `reset` mode, but withouht the bottom `cancel` button action to go back to `signin`.  This method is useful when your site has custom *signin* and *signup* links at a different form.

You may call this method with a single parameter, two or even none. The following examples ilustrate this:

```js
var lock = new Auth0Lock('xxxxxx', '<account>.auth0.com');

// using defaults and resolved configuration
// from you account details
lock.showReset();

// passing some options
lock.showReset(options);

// provide with a callback `fn` to be invoked
// on `reset` success or error
lock.showReset(function (err) {});

// or both options and callback
lock.showReset(options, function (err) {});
```
> Check the [Auth0Lock customization][lock-customization] article for more examples and options specification. Or enter the [Authentication modes][application-types] notes to learn more about implementing different authentication mechanics.


### .hide([callback])

Close the widget and invoke `callback` when removed from DOM.

```js
var lock = new Auth0Lock('xxxxxx', '<account>.auth0.com');

// normal display
lock.show(options);

// trigger hide when esc key pressed
document.addEventListener('keypress', function(e) {
  // hide if esc
  lock.hide();
}, false);
```

### .logout([query])

Log out loggedin user with optional query parameters for the `GET` request.

```js
var lock = new Auth0Lock('xxxxxx', '<account>.auth0.com');

// Call logout with query parameters
lock.logout({ ref: window.location.href });
```

## Examples

The **example** directory has a ready-to-go app. In order to run it you need [node](http://nodejs.org/) installed.

Then execute `npm i` to install dependencies (only once) and `npm example` from the root of this project.

Finally, point your browser at `http://localhost:3000/` and play around.

## Browser Compatibility

We ensure browser compatibility in `Chrome`, `Safari`, `Firefox` and `IE >= 9`. We currently use [zuul](https://github.com/defunctzombie/zuul) along with [Saucelabs](https://saucelabs.com) to run integration tests on each push.

## Issue Reporting

If you have found a bug or if you have a feature request, please report them at this repository issues section. Please do not report security vulnerabilities on the public GitHub issue tracker. The [Responsible Disclosure Program](https://auth0.com/whitehat) details the procedure for disclosing security issues.

## Resources

* [Complete API][lock-customization]
* [UI customization][ui-customization]
* [Single Page Applications][spa-notes] implementation notes.
* [Regular Web Applications][webapps-notes] implementing notes.
* [Overlay vs Embedded mode][display-modes]
* [Popup vs Redirect mode][authentication-modes] notes. **What are the authentication modes?**.
* [Error customization][error-customization] notes.
* [I18n][i18n-notes] notes.
* [Events][events-notes] notes.
* [Development][development-notes] notes.
* [Release process][release-process] notes.
* [Auth0Lock playground][playground-url]
* [Using Refresh Tokens](https://github.com/auth0/lock/wiki/Using-a-Refresh-Token)
* Legacy **Auth0Widget**`< v6.0.0` [repository](https://github.com/auth0/widget) and [docs](https://docs.auth0.com/login-widget2).
* Legacy **Auth0Widget** [Migration guide][migration-guide] to **AuthoLock**

<!-- Vaaaaarrsss -->

[download1]: https://raw.github.com/auth0/lock/master/build/auth0-lock.js
[download2]: https://raw.github.com/auth0/lock/master/build/auth0-lock.min.js

[npm-image]: https://img.shields.io/npm/v/auth0-lock.svg?style=flat-square
[npm-url]: https://npmjs.org/package/auth0-lock
[strider-image]: https://ci.auth0.com/auth0/lock/badge
[strider-url]: https://ci.auth0.com/auth0/lock
[coveralls-image]: https://img.shields.io/coveralls/auth0/lock.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/auth0/lock?branch=master
[david-image]: http://img.shields.io/david/auth0/lock.svg?style=flat-square
[david-url]: https://david-dm.org/auth0/lock
[license-image]: http://img.shields.io/npm/l/auth0-lock.svg?style=flat-square
[license-url]: LICENSE
[downloads-image]: http://img.shields.io/npm/dm/auth0-lock.svg?style=flat-square
[downloads-url]: https://npmjs.org/package/auth0-lock

[lock-initialization]: https://github.com/auth0/lock/wiki/Auth0lock-initialization
[lock-customization]: https://github.com/auth0/lock/wiki/Auth0lock-customization
[application-types]: https://github.com/auth0/lock/wiki/Types-Of-Applications
[display-modes]: https://github.com/auth0/lock/wiki/Display-Modes
[spa-notes]: https://github.com/auth0/lock/wiki/Types-Of-Applications#single-page-app
[webapps-notes]: https://github.com/auth0/lock/wiki/Types-Of-Applications#regular-webapp
[authentication-modes]: https://github.com/auth0/lock/wiki/Authentication-Modes
[popup-mode]: https://github.com/auth0/lock/wiki/Authentication-Modes#popup-mode
[redirect-mode]: https://github.com/auth0/lock/wiki/Authentication-Modes#redirect-mode
[ui-customization]: https://github.com/auth0/lock/wiki/UI-customization
[error-customization]: https://github.com/auth0/lock/wiki/Customizing-error-messages
[i18n-notes]: https://github.com/auth0/lock/wiki/I18n
[events-notes]: https://github.com/auth0/lock/wiki/Events
[development-notes]: https://github.com/auth0/lock/wiki/Development-notes
[release-process]: https://github.com/auth0/lock/wiki/Release-process
[playground-url]: http://auth0.github.com/playground
[migration-guide]: https://github.com/auth0/lock/wiki/Migration-guide


History.md
## 7.1.5 - (Feb 25, 2015)

 - [lock] fix(i18n): fix some minor german translation inconsistencies (`gjungb`)
  https://github.com/auth0/lock/commit/5aa6cd9284e803d158071a8575bf4fa9359e8449

## 7.1.4 - (Feb 24, 2015)

### Fixed

 - [lock] Fixes #129: /ro mode not working when called with popup: true and sso: false (`pose`)
  https://github.com/auth0/lock/commit/c80972a89f32a8ea358dd90d54e09c0b7eac2c49

## 7.1.3 - (Feb 17, 2015)

### Changed

 - [lock] Updating to auth0.js 6.0.6:
   * #112: Fixes problems with Phonegap and db connections.
   * If dbconn and callback.length <1 arguments we don't open a popup.
   * Adding popup blocker warning. (`pose`)

  https://github.com/auth0/lock/commit/0e0d726ca58e1119b8914cfadb9ebf1be99741f4

## 7.1.2 - (Feb 17, 2015)

### Fixed

 - [lock] Fixes #122: Correct german translations. (`pose`)
  https://github.com/auth0/lock/commit/9bb2546f2669eef78c74903ce9fca7088cf9dabe

## 7.1.1 - (Feb 2, 2015)

### Added

 - [lock] Update fr-FR.json (Updated to reflect the lastest en.json): When default language is switched to french, it causes an issue with ionic automatic generated sample downloadable from auth0.com, and probably with any configuration using Lock : (`bonatoc`)
  https://github.com/auth0/lock/commit/6cc8bdf358fd660e2b641a011b2d342e3dc29362

### Fixed

 - [lock] Fixes #115: Wrong error message on sign up with sso domain. (`pose`)
  https://github.com/auth0/lock/commit/4d25f3257a97a0fe041c75fadbc3f3a7be4da329

## 7.1.0 - (Feb 2, 2015) [YANKED]

## 7.0.0 - (Jan 27, 2015)

### Changed

 - [lock] Now sso: true works for real (`pose`)
  https://github.com/auth0/lock/commit/525c1f41f9a1df57baf84b6749cbf2629a8c7915
 - [lock] Updating to auth0.js 6.0.2 (`pose`)
  https://github.com/auth0/lock/commit/ece664925a2643ee6be274d353309d1a32c3ed9f

## 6.12.1 - (Jan 23, 2015)

### Changed

 - [lock] Updating to auth0-js 5.5.1 (`pose`)
  https://github.com/auth0/lock/commit/d0f2d1916963dfccd977017d5cf9418a491309f8

## 6.12.0 - (Jan 23, 2015)

### Added

 - [lock] Add flag to enable cordova plugin auth for Lock. (`hzalaz`)
  https://github.com/auth0/lock/commit/cf0a7c07a9a84da432da416e177ae92cb4246ad0

## 6.11.0 - (Jan 23, 2015)

### Changed

 - [lock] Updating to auth0.js 5.5.0 (`pose`)
  https://github.com/auth0/lock/commit/b4da2345ed2827d13300a0d8f42d85ea0d85d68a

## 6.10.6 - (Jan 22, 2015)

### Fixed

 - [lock] Improving iOS detection. (`pose`)
  https://github.com/auth0/lock/commit/7ea0e77cce25fce09bd09e01537d47cecb81b66e

## 6.10.5 - (Jan 22, 2015)

### Fixed

 - [lock] Fixes #107: Lock looks bad on iPhone. (`pose`)
  https://github.com/auth0/lock/commit/b0403e3a177129cc22136af131c4bd080362ef68

## 6.10.4 - (Jan 20, 2015)

### Added

 - [lock] add exact as strategy (`siacomuzzi`)
  https://github.com/auth0/lock/commit/618fe469e3d4a3448664c6960e0bb59268e82011

## 6.10.3 - (Jan 20, 2015)

### Added

 - [lock] Add "exact" social provider icon to Lock (`vctrfrnndz`)
  https://github.com/auth0/lock/commit/a7ea4f0c3f7212c17feb370313b461719d8c38e5

### Fixed

 - [lock] Remove ttf data-uri to reduce file size (`vctrfrnndz`)
  https://github.com/auth0/lock/commit/a9a46fa05eb0a9cfbafe9abb19601f530b963638

## 6.10.2 - (Jan 19, 2015)

### Fixed

 - [lock] fix build for windows (`jfromaniello`)
  https://github.com/auth0/lock/commit/52bcd01b6c7e71aa82815365634dd7a3289d3919


## 6.10.1 - (Jan 16, 2015)

### Added

 - [lock] Update pt-BR.json (`gabrielmoreira`)
  https://github.com/auth0/lock/commit/9884061ddf71ba3ca4a9d58109802e9f4b1e3497
 - [lock] Better message for identicalChars (`gabrielmoreira`)
  https://github.com/auth0/lock/commit/65ef017d381d7babd835408438c000205bdc18d3
 - [lock] Smaller social button text for a better UI (`gabrielmoreira`)
  https://github.com/auth0/lock/commit/9efe72537abcc07412d321cb39c3afc4870bddde
 - [lock] added danish translation (`carnevalle`)
  https://github.com/auth0/lock/commit/1b7d05f9bb2bbf10789218515213faaa4412ebd3
 - [lock] Fix path for danish translations (`cristiandouce`)
  https://github.com/auth0/lock/commit/64a77404874f50aeec0f4831cc015268c3b077fc
 - [lock] Added da.json to dics_data in lib/i18n/index.js (`carnevalle`)
  https://github.com/auth0/lock/commit/18b65bbc04cdf5e191dd039bf6629d6d4080486e
 - [lock] Add support for "signing starting" event (`sandrinodimattia`)
  https://github.com/auth0/lock/commit/49fe7c650cadf231850498b05f96ac6d5fbd00ed
 - [lock] Renamed event 'signing starting' to 'signin' + test (`sandrinodimattia`)
  https://github.com/auth0/lock/commit/d2b387b6a19d4fff2b1986845f5bf1edfed57b79
 - [lock] Make signin event more verbose. Renamed to signin submit. (`sandrinodimattia`)
  https://github.com/auth0/lock/commit/b191cd026f796f69f87396b694d04df06694e9de

### Fixed

 - [lock] remove trailing commas to improve compatibility (`nelix`)
  https://github.com/auth0/lock/commit/20876e56aecedb3a2bde6530e3038a98c0bb6642
 - [lock] Fixes check for console by doing window.close (`pose`)
  https://github.com/auth0/lock/commit/adf0ad269f2980f951bb01bf592fd1dd3e15ce20

### Changed

 - [lock] Using auth0.js version 5.2.2 (`pose`)
  https://github.com/auth0/lock/commit/6b4f5ed8efc7b3e8ebb3a1c1d33b40f474d84e96


## 6.10.0 - (Jan 16, 2015) [YANKED]

## 6.9.0 - (Jan 16, 2015) [YANKED]

## 6.8.4 - (Jan 8, 2015) [YANKED]

## 6.8.3 - (Jan 6, 2015) [YANKED]

## 6.8.2 - (Jan 2, 2015) [YANKED]

## 6.8.1 - (Jan 2, 2015) [YANKED]

## 6.8.0 - (Dec 31, 2014) [YANKED]

## 6.7.1 - (Dec 30, 2014) [YANKED]

## 6.7.0 - (Dec 24, 2014) [YANKED]

## 6.6.5 - (Dec 15, 2014)

### Fixed

 - [lock] remove duplicate keys which break json5

This relates to #54, because most of the other translations require json5-loader for this project to be built with webpack (browserify alternative).
Another idea would be to fix the trailing commas in some of the other translations so they could be loaded with json-loader (which supports duplicate keys). (`nelix`)
  https://github.com/auth0/lock/commit/72d5c63ad12a18f019a8deeadd7f3bb6108c97d2

## 6.6.4 - (Dec 15, 2014)


Gruntfile.js
var fs = require('fs');
var path = require('path');
var pkg = require('./package');
var cssPrefix = require('css-prefix');

var minor_version = pkg.version.replace(/\.(\d)*$/, '');
var major_version = pkg.version.replace(/\.(\d)*\.(\d)*$/, '');
var path = require('path');

function  rename_release (v) {
  return function (d, f) {
    var dest = path.join(d, f.replace(/(\.min)?\.js$/, '-'+ v + '$1.js').replace('auth0-', ''));
    return dest;
  };
}

function node_bin (bin) {
  return path.join('node_modules', '.bin', bin);
}

module.exports = function (grunt) {
  grunt.initConfig({
    connect: {
      test: {
        options: {
          // base: 'test',
          hostname: '*',
          base: ['.', 'example', 'example/build', 'build'],
          port: 9999
        }
      },
      example: {
        options: {
          hostname: '*',
          base: ['example', 'example/build', 'build'],
          port: 3000
        }
      },
      'example-https': {
        options: {
          base: ['example', 'example/build', 'build'],
          port:  3000,
          protocol: 'https',
          hostname: '*',
          cert: fs.readFileSync(__dirname + '/test/support/https-certs/server.crt').toString(),
          key:  fs.readFileSync(__dirname + '/test/support/https-certs/server.key').toString()
        }
      }
    },
    browserify: {
      options: {
        browserifyOptions: {
          debug: true
        },
        watch: true,

        // Convert absolute sourcemap filepaths to relative ones using mold-source-map.
        postBundleCB: function(err, src, cb) {
          if (err) { return cb(err); }
          var through = require('through');
          var stream = through().pause().queue(src).end();
          var buffer = '';

          stream.pipe(require('mold-source-map').transformSourcesRelativeTo(__dirname)).pipe(through(function(chunk) {
            buffer += chunk.toString();
          }, function() {
            cb(err, buffer);
          }));
          stream.resume();
        }

      },
      debug: {
        files: {
          'build/auth0-lock.js': ['standalone.js']
        }
      },
    },
    less: {
      dist: {
        options: {
          paths: ['lib/css'],
        },
        files: {
          'lib/css/main.css': 'lib/css/main.less'
        }
      },
      example: {
        files: {
          'example/build/index.css': 'example/index.less'
        }
      }
    },
    autoprefixer: {
      options: {
        browsers: ['> 1%', 'last 2 versions', 'ff 15', 'opera 12.1', 'ie 8']
      },
      main: {
        src:  'lib/css/main.css',
        dest: 'lib/css/main.css',
      },
    },
    prefix: { //this adds 'a0-' to every class and id
      css: {
        src: 'lib/css/main.css',
        dest: 'lib/css/main.css',
        prefix: 'a0-'
      }
    },
    cssmin: {
      minify: {
        options: {
          keepSpecialComments: 0
        },
        files: {
          'lib/css/main.min.css': ['lib/css/main.css']
        }
      }
    },
    copy: {
      example: {
        files: {
          'example/auth0-lock.min.js': 'build/auth0-lock.min.js',
          'example/auth0-lock.js':     'build/auth0-lock.js'
        }
      },
      release: {
        files: [
          { expand: true, flatten: true, src: 'build/*', dest: 'release/', rename: rename_release(pkg.version) },
          { expand: true, flatten: true, src: 'build/*', dest: 'release/', rename: rename_release(minor_version) },
          { expand: true, flatten: true, src: 'build/*', dest: 'release/', rename: rename_release(major_version) }
        ]
      }
    },
    exec: {
      'uglify': {
        cmd: node_bin('uglifyjs') + ' build/auth0-lock.js  -b beautify=false,ascii_only=true > build/auth0-lock.min.js',
        stdout: true,
        stderr: true
      },
      'test-inception': {
        cmd: node_bin('mocha') + ' ./test/support/characters-inception.test.js',
        stdout: true,
        stderr: true
      },
      'test-integration': {
        cmd: node_bin('zuul') + ' -- test/*.js',
        stdout: true,
        stderr: true
      },
      'test-phantom': {
        cmd: node_bin('zuul') + ' --ui mocha-bdd --phantom 9999 -- test/*.js',
        stdout: true,
        stderr: true
      }
    },
    clean: {
      css: ['lib/css/main.css', 'lib/css/main.min.css'],
      js: ['release/', 'build/', 'example/auth0-lock.js']
    },
    watch: {
      js: {
        files: ['build/auth0-lock.js'],
        tasks: [],
        options: {
          livereload: true
        },
      },
      css: {
        files: [
          'lib/**/*.less'
        ],
        tasks: ['build'],
        options: {
          livereload: true
        },
      },
      example: {
        files: ['example/*'],
        tasks: ['less:example'],
        options: {
          livereload: true
        },
      }
    },
    aws_s3: {
      options: {
        accessKeyId:     process.env.S3_KEY,
        secretAccessKey: process.env.S3_SECRET,
        bucket:          process.env.S3_BUCKET,
        uploadConcurrency: 5,
        params: {
          CacheControl: 'public, max-age=300'
        },
        // debug: true <<< use this option to test changes
      },
      clean: {
        files: [
          { action: 'delete', dest: 'js/lock-' + pkg.version + '.js' },
          { action: 'delete', dest: 'js/lock-' + pkg.version + '.min.js' },
          { action: 'delete', dest: 'js/lock-' + major_version + '.js' },
          { action: 'delete', dest: 'js/lock-' + major_version + '.min.js' },
          { action: 'delete', dest: 'js/lock-' + minor_version + '.js' },
          { action: 'delete', dest: 'js/lock-' + minor_version + '.min.js' }
        ]
      },
      publish: {
        files: [
          {
            expand: true,
            cwd:    'release/',
            src:    ['**'],
            dest:   'js/'
          }
        ]
      }
    },
    /* Checks for outdated npm dependencies before release. */
    outdated: {
      release: {
        development: false
      }
    },
    /* Purge FASTLY cache. */
    fastly: {
      options: {
        key:  process.env.FASTLY_KEY,
        host: process.env.FASTLY_HOST
      },
      purge: {
        options: {
          urls: [
            'js/lock-' + pkg.version + '.js',
            'js/lock-' + pkg.version + '.min.js',
            'js/lock-' + major_version + '.js',
            'js/lock-' + major_version + '.min.js',
            'js/lock-' + minor_version + '.js',
            'js/lock-' + minor_version + '.min.js',
          ]
        },
      },
    },
    http: {
      purge_js: {
        options: {
          url: process.env.CDN_ROOT + '/js/lock-' + pkg.version + '.js',
          method: 'DELETE'
        }
      },
      purge_js_min: {
        options: {
          url: process.env.CDN_ROOT + '/js/lock-' + pkg.version + '.min.js',
          method: 'DELETE'
        }
      },
      purge_major_js: {
        options: {
          url: process.env.CDN_ROOT + '/js/lock-' + major_version + '.js',
          method: 'DELETE'
        }
      },
      purge_major_js_min: {
        options: {
          url: process.env.CDN_ROOT + '/js/lock-' + major_version + '.min.js',
          method: 'DELETE'
        }
      },
      purge_minor_js: {
        options: {
          url: process.env.CDN_ROOT + '/js/lock-' + minor_version + '.js',
          method: 'DELETE'
        }
      },
      purge_minor_js_min: {
        options: {
          url: process.env.CDN_ROOT + '/js/lock-' + minor_version + '.min.js',
          method: 'DELETE'
        }
      }
    }
  });

  grunt.registerMultiTask('prefix', 'Prefix css.', function() {
    var done = this.async();
    var that = this;
    fs.readFile(__dirname + '/' + this.data.src, {encoding: 'utf8'}, function (err, data) {
      if (err) { return done(err);  }
      var prefixed = cssPrefix(that.data.prefix, data.toString());
      fs.writeFile(__dirname + '/' + that.data.dest, prefixed, function (err) {
        if (err) { return done(err); }
        done();
      });
    });
  });

  // Loading dependencies
  for (var key in grunt.file.readJSON('package.json').devDependencies) {
    if (key !== 'grunt' && key.indexOf('grunt') === 0) { grunt.loadNpmTasks(key); }
  }

  grunt.registerTask('css',           ['clean:css', 'less:dist', 'prefix:css', 'autoprefixer:main', 'cssmin:minify']);

  grunt.registerTask('js',            ['clean:js', 'browserify:debug', 'exec:uglify']);
  grunt.registerTask('build',         ['css', 'js']);

  grunt.registerTask('example',       ['less:example', 'connect:example', 'build', 'watch']);
  grunt.registerTask('example-https', ['less:example', 'connect:example-https', 'build', 'watch']);

  grunt.registerTask('dev',           ['connect:test', 'build', 'watch']);
  grunt.registerTask('integration',   ['exec:test-inception', 'exec:test-integration']);
  grunt.registerTask('phantom',       ['build', 'exec:test-inception', 'exec:test-phantom']);

  grunt.registerTask('purge_cdn',     ['http:purge_js', 'http:purge_js_min', 'http:purge_major_js', 'http:purge_major_js_min', 'http:purge_minor_js', 'http:purge_minor_js_min']);

  grunt.registerTask('cdn',           ['build', 'copy:release', 'aws_s3:clean', 'aws_s3:publish', 'purge_cdn']);
};
